src/sage/combinat/integer_vectors_mod_permgroup.py

```diff
@@ -775,6 +785,79 @@ def __iter__(self):
             else:
                 return SF.elements_of_depth_iterator(self._sum)
 
+    def cardinality(self):
+        r"""Return the number of integer vectors in the set.
+
+        The number is computed using the cycle index theorem, which is
+        faster than listing the vectors.
+
+        EXAMPLES::
+
+            sage: G = PermutationGroup([], domain=[1,2,3])
+            sage: IntegerVectorsModPermutationGroup(G, 5).cardinality()
+            21
+
+            sage: G = PermutationGroup([(1,2)])
+            sage: IntegerVectorsModPermutationGroup(G, 1000).cardinality()
+            501
+
+            sage: G = PermutationGroup([(1,2,3)])
+            sage: I = IntegerVectorsModPermutationGroup(G, sum=10, max_part=5)
+            sage: I.cardinality()
+            7
+
+            sage: G = SymmetricGroup(10)
+            sage: I = IntegerVectorsModPermutationGroup(G, max_part=1)
+            sage: I.cardinality()
+            11
+        """
+        G = self._permgroup
+        k = G.degree()          # Vector length
+        d = self._sum           # Required total
+        m = self._max_part      # Max of one entry, -1 for no limit
+        if m == -1:
+            m = d               # Any entry cannot exceed total
+        if k == 0:
```

Comment:
alternatively: `if not k:`

Reply:
Umm... I'd rather not. We are really talking about integers here, I think it is much more natural to compare to zero than to treat the integer as a pseudo Boolean. Is there an established rule for using ```not k``` pro ```k == 0``` when ```k``` is really and conceptually an integer value?

Reply:
It is faster and more common (at least within Sage's library) to do this. Most integers end up not being Python `int`'s as well, meaning coercion gets involved.

Reply:
Faster my left foot. The comparison is done once (as in, "1 time") per invocation of `cardinality`. The difference is what, 40 nanoseconds? Did you measure?
```
sage: V=IntegerVectorsModPermutationGroup(SymmetricGroup(5),100)
sage: timeit("V.fast_cardinality()", number=1000, precision=5)
1000 loops, best of 3: 2.4201 ms per loop
sage: timeit("V.faster_cardinality()", number=1000, precision=5)
1000 loops, best of 3: 2.4255 ms per loop
sage: timeit("V.cardinality()", number=1, precision=5)
1 loop, best of 3: 10.837 s per loop
```
Here `cardinality` is the old iterative version. `fast_cardinality` is the version with `k==0` and `faster_cardinality` is exactly same code with `not k`. Perhaps the `not k` version is faster by a few nanoseconds, but I am not seeing it here. We are replacing an old 10-second algorithm with a new 2.4-millisecond algorithm, and are we now seriously talking about shaving nanoseconds?

Since when did SageMath programmers become interested in shaving nanoseconds anyway? Quoting 
https://doc.sagemath.org/html/en/faq/faq-contribute.html
> Fast code is valuable, but clean, readable code is important. In the mathematics community, inaccurate results are unacceptable.  ... Don Knuth observes that: “We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil.”

Replacing the clean, semantically transparent integer comparison `k==0` with a pseudo-Boolean comparison `not k` is, in my humble opinion, a prime example of a futile attempt at premature optimization that obscures the meaning and works as an open invitation to bugs. And it is against the FAQ that I quoted above.

I ask again: Is this `not k` style for integer comparison **endorsed in some official documentation**, or is it just a habit that SageMath programmers have grown fond of, for better or worse?

Reply:
For the record, yes, I have repeatedly measured. Not every coding convention (in fact, many of them are not) is in some official documentation. Not everything is called "1 time" such as the method being called in a tight loop. This is also not premature optimization but a standard micro-optimization; premature optimization is writing very complicated code for something that is not a bottleneck (as one has not profiled code). For example, another oddity of Python is `tuple([i for i in range(5)])` is faster than `tuple(i for i in range(5))`, so we often use the former even though the latter is more "clean" (in some sense of the word).

I hope you will continue to want to be an active and collaborative part of the SageMath developers community. Yet, I feel I must also point out our [code of conduct](https://github.com/sagemath/sage/blob/develop/CODE_OF_CONDUCT.md) and that your tone is close to not adhering to those standards.

Reply:
Well, if that thing is a "standard micro-optimization" routinely applied in SageMath code, I must say I am sorry to hear that.

With all due respect I disagree that it is a good idea to apply it all over the place, or here. I have expressed the reasons for my stance and I will leave it that.

I will *not* be applying that kind of thing here, but if someone else wants to, then that's their right. I will leave it at that.

Reply:
Dear @jukkakohonen!

I am very sorry that this got so heated, and I would like to ask you to stay.  I cannot speak for Travis, but I think what happened is a mis-emphasis.  Initially, that is, when I switched to sage and therefore python, I was also extremely annoyed by some (and in fact, the sheer number of) idioms.  (My background is common lisp and later, Aldor, which are - in my opinion - very clean languages.  Lisp doesn't even have much syntax.)

In the case at hand, I think it is mostly a matter of coding convention.  Some of these have been written down, some of these have just formed over time.  I don't know whether `not x` instead of `x == 0` is written down somewhere, but it is certainly standard, because in a tight loop you want to choose the faster version, and if I don't read `not x` over and over again, I might not realize that it is a way to check whether `x` is zero.  I remark that it is not applied always: grep finds about 6000 matches of ` == 0` in sage library code.  Of course, many of those are leftovers of past times.

In any case, I like your contribution a lot, and find it very helpful.  So please, let's get it done and let's focus on what we can agree on.

Reply:
So, could you simply reshape the pull request otherwise as discussed (separating the `d is None` case, perhaps rename the variables to make their meaning even more obvious, and perhaps adding a systematic test, and if not, at least one other non-cyclic, non-symmetric group and the test showing that #36681 is fixed)?